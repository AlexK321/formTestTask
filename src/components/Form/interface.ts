export interface FormData {
  email?: string;
  phoneNumber?: string;
  userName?: string;
}
