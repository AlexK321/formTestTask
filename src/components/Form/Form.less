form {
  background-color: white;
  border-radius: 4px;
  padding: 24px 40px!important;
  width: 100%;
  input {
    border-radius: 4px!important;
    padding: 25px 18px 4px 11px!important;
  }
  button {
    border: 1px solid #00b4e3!important;
    background-color: #00b4e3!important;
    border-radius: 8px!important;
    color: #fff!important;
    padding: 0 50px!important;
    height: 60px!important;
    &:hover {
      border-color: #17cfff!important;
      background-color: #17cfff!important;
    }
  }
}  