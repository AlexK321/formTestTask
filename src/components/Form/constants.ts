export const USER_NAME = 'userName';
export const PHONE_NUMBER = 'phoneNumber';
export const EMAIL = 'email';
export const POST_URL = 'https://jsonplaceholder.typicode.com/posts';
export const LABEL_COL = { span: 8 };
export const WRAPPER_COL = { span: 16 };
export const FILLED_SEARCH_QUERY = 'filled';
export const CORRECT_SMS_CODE = '1111';
export const TIME_DELAY = 20;
export const MILLISECONDS = 1000;
export const MAX_SMS_CODE_ATTEMPTS = 3;
