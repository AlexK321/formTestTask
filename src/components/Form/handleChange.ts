const handleChange = (formData: FormData): void => {
  const formDataEntries = Object.entries(formData);
  const [key, value] = formDataEntries[0];

  localStorage.setItem(key, value);
};

export default handleChange;
