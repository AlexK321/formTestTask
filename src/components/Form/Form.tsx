import React, { ReactElement, useState } from 'react';
import { Form as FormComponent, Button } from 'antd';
import './Form.css';
import axios from 'axios';
import SubmitResult from '../SubmitResult/SubmitResult';
import FormItem from '../FormItem/FormItem';

interface FormData {
  email?: string;
  phoneNumber?: string;
  userName?: string;
}

const userName = 'userName';
const phoneNumber = 'phoneNumber';
const email = 'email';
const labelCol = { span: 8 };
const wrapperCol = { span: 16 };

const initialValue = {
  userName: localStorage.getItem(userName) || '',
  phoneNumber: localStorage.getItem(phoneNumber) || '',
  email: localStorage.getItem(email) || '',
};

const changeInitialValue = (initialValue1: FormData, filled1: string | null) => {
  const initialValueArr = Object.entries(initialValue1);
  for (let i = Number(filled1); i < initialValueArr.length; i += 1) {
    initialValueArr[i][1] = '';
  }
  return Object.fromEntries(initialValueArr);
};

const getInitialValuesForm = () => {
  const url = new URL(window.location.href);
  const filled = url.searchParams.get('filled');
  if (filled === '0') {
    url.searchParams.delete('filled');
    document.location.href = url.toString();
  } else if (Number(filled) > 0) {
    return changeInitialValue(initialValue, filled);
  }
  return initialValue;
};

const Form = (): ReactElement => {
  const [requestStatus, setRequestStatus] = useState<number | null>(null);

  const sendFormData = async (formData: FormData) => {
    try {
      const response = await axios.post('https://jsonplaceholder.typicode.com/posts', formData);
      setRequestStatus(response.status);
    } catch (err) {
      console.error(err);
    }
  };

  const handleChange = (formData: FormData): void => {
    const formDataArray = Object.entries(formData);
    localStorage.setItem(String(formDataArray[0][0]), String(formDataArray[0][1]));
  };

  if (requestStatus) return <SubmitResult status={requestStatus} />;

  return (
    <FormComponent
      labelCol={labelCol}
      wrapperCol={wrapperCol}
      initialValues={getInitialValuesForm()}
      onFinish={sendFormData}
      onValuesChange={handleChange}
    >
      <h2>Контакты</h2>
      <FormItem name="userName" placeholder="Ваше имя" />
      <div className="form-row">
        <FormItem name="phoneNumber" placeholder="Телефон" />
        <FormItem name="email" placeholder="E-mail" />
      </div>
      <div className="form-row">
        <Button type="primary" htmlType="submit">
          Отправить заявку
        </Button>
        <p>
          Нажимая на кнопку «Отправить заявку», вы соглашаетесь на обработку персональных данных
        </p>
      </div>
    </FormComponent>
  );
};

export default Form;
