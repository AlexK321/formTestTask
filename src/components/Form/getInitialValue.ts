import { FILLED } from './constants';

interface FormData {
  email?: string;
  phoneNumber?: string;
  userName?: string;
}

export const changeInitialValue = (formData: FormData, filled: number): FormData => {
  const filteredFormData = Object.entries(formData).slice(0, filled);

  return Object.fromEntries(filteredFormData);
};

const getInitialValues = (formData: FormData): FormData => {
  let finalFormData = formData;
  const url = new URL(window.location.href);
  const filled = Number(url.searchParams.get(FILLED));
  const maxItems = Object.entries(formData).length;

  if (filled > 0 && filled <= maxItems) {
    finalFormData = changeInitialValue(formData, filled);
  } else if (url.searchParams.has(FILLED)) {
    url.searchParams.delete(FILLED);
    window.location.href = url.toString();
  }

  return finalFormData;
};

export default getInitialValues;
