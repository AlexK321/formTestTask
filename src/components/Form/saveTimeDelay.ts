import { TIME_DELAY } from './constants';

const saveTimeDelay = (): void => {
  const date: Date = new Date();
  const finishDate = date.setSeconds(date.getSeconds() + TIME_DELAY);

  localStorage.setItem('finishDate', JSON.stringify(finishDate));
};

export default saveTimeDelay;
