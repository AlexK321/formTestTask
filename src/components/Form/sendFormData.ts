import React from 'react';
import axios from 'axios';
import { POST_URL } from './constants';
import { FormData } from './interface';

const sendFormData = async (
  formData: FormData,
  setHasError: React.Dispatch<React.SetStateAction<boolean | null>>
): Promise<void> => {
  try {
    await axios.post(POST_URL, formData);
    setHasError(false);
  } catch (err) {
    setHasError(true);
  }
};

export default sendFormData;
