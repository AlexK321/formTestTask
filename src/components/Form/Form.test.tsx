import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import axios from 'axios';
import Form from './Form';
import SubmitResult from '../SubmitResult/SubmitResult';

jest.mock('axios');

describe('Component: Form', () => {
  let component: ShallowWrapper;

  beforeEach(() => {
    component = shallow(<Form />);
  });

  describe('when request status exists', () => {
    const status = 200;

    beforeEach(() => {
      (axios.post as jest.Mock).mockResolvedValue({ status });
    });

    describe('when form is finished with result ', () => {
      beforeEach(() => {
        component.simulate('finish');
      });

      it('should render SubmitResult', () => {
        const submitResult = component.find(SubmitResult);

        expect(submitResult).toHaveLength(1);
        expect(submitResult.props()).toStrictEqual({ hasError: false });
      });
    });
  });

  describe('when reject status exists', () => {
    const status = 404;

    beforeEach(() => {
      (axios.post as jest.Mock).mockRejectedValue({ status });
      component.update();
    });

    describe('when form is finished with result ', () => {
      beforeEach(() => {
        component.simulate('finish');
      });

      it('should render SubmitResult', () => {
        const submitResult = component.find(SubmitResult);

        expect(submitResult).toHaveLength(1);
        expect(submitResult.props()).toStrictEqual({ hasError: true });
      });
    });
  });

  it('should render title', () => {
    const consentDataProcessing =
      'Нажимая на кнопку «Отправить заявку», вы соглашаетесь на обработку персональных данных';
    const wrapper = component.find('.form-description');

    expect(wrapper.text()).toBe(consentDataProcessing);
  });

  it('should render button', () => {
    const wrapper = component.find('Button');

    expect(wrapper).not.toBeNull();
  });

  describe('when change input', () => {
    beforeEach(() => {
      component.simulate('valuesChange', {
        email: 'e',
      });
    });

    it('should change localStorage', () => {
      expect(localStorage.setItem).toHaveBeenCalledTimes(1);
    });
  });
});
