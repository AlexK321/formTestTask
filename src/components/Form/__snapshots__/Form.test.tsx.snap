// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Form snapshot component 1`] = `
<ForwardRef(InternalForm)
  initialValues={
    Object {
      "email": "",
      "phoneNumber": "",
      "userName": "",
    }
  }
  labelCol={
    Object {
      "span": 8,
    }
  }
  onFinish={[Function]}
  onValuesChange={[Function]}
  wrapperCol={
    Object {
      "span": 16,
    }
  }
>
  <h2>
    Контакты
  </h2>
  <FormItem
    name="userName"
    placeholder="Ваше имя"
  />
  <div
    className="form-row"
  >
    <FormItem
      name="phoneNumber"
      placeholder="Телефон"
    />
    <FormItem
      name="email"
      placeholder="E-mail"
    />
  </div>
  <div
    className="form-row"
  >
    <Button
      htmlType="submit"
      type="primary"
    >
      Отправить заявку
    </Button>
    <p>
      Нажимая на кнопку «Отправить заявку», вы соглашаетесь на обработку персональных данных
    </p>
  </div>
</ForwardRef(InternalForm)>
`;

exports[`Component: Form when render elements snapshot component 1`] = `
<ForwardRef(InternalForm)
  initialValues={
    Object {
      "email": "",
      "phoneNumber": "",
      "userName": "",
    }
  }
  labelCol={
    Object {
      "span": 8,
    }
  }
  onFinish={[Function]}
  onValuesChange={[Function]}
  wrapperCol={
    Object {
      "span": 16,
    }
  }
>
  <h2>
    Контакты
  </h2>
  <FormItem
    name="userName"
    placeholder="Ваше имя"
  />
  <div
    className="form-row"
  >
    <FormItem
      name="phoneNumber"
      placeholder="Телефон"
    />
    <FormItem
      name="email"
      placeholder="E-mail"
    />
  </div>
  <div
    className="form-row"
  >
    <Button
      htmlType="submit"
      type="primary"
    >
      Отправить заявку
    </Button>
    <p>
      Нажимая на кнопку «Отправить заявку», вы соглашаетесь на обработку персональных данных
    </p>
  </div>
</ForwardRef(InternalForm)>
`;

exports[`Component: Form when request status is null should match snapshot 1`] = `
<ForwardRef(InternalForm)
  initialValues={
    Object {
      "email": "",
      "phoneNumber": "",
      "userName": "",
    }
  }
  labelCol={
    Object {
      "span": 8,
    }
  }
  onFinish={[Function]}
  onValuesChange={[Function]}
  wrapperCol={
    Object {
      "span": 16,
    }
  }
>
  <h2>
    Контакты
  </h2>
  <FormItem
    name="userName"
    placeholder="Ваше имя"
  />
  <div
    className="form-row"
  >
    <FormItem
      name="phoneNumber"
      placeholder="Телефон"
    />
    <FormItem
      name="email"
      placeholder="E-mail"
    />
  </div>
  <div
    className="form-row"
  >
    <Button
      htmlType="submit"
      type="primary"
    >
      Отправить заявку
    </Button>
    <p>
      Нажимая на кнопку «Отправить заявку», вы соглашаетесь на обработку персональных данных
    </p>
  </div>
</ForwardRef(InternalForm)>
`;
